mlflow/types/utils.py

```diff
@@ -337,42 +372,34 @@ def _infer_numpy_dtype(dtype) -> DataType:
     raise MlflowException(f"Unsupported numpy data type '{dtype}', kind '{dtype.kind}'")
 
 
+def _is_none_or_nan(x):
+    if isinstance(x, float):
+        return np.isnan(x)
+    return x is None
+
+
+def _infer_required(col) -> bool:
+    if hasattr(col, "__iter__"):
+        return not any(_is_none_or_nan(x) for x in col)
+    return not _is_none_or_nan(col)
```

Comment:
Can `col` be a string?

Reply:
Yes. If it's an empty string "", I think we should still infer it as a string type.

Reply:
Can `col` be a dict?

Reply:
Yes. Empty dictionary is also acceptable.

Reply:
Can we use `instance(col, list)`? `hasattr(col, "__iter__")` matches many objects.

Reply:
Yea will update it to `isinstance(col, (list, pd.Series))` in my follow-up PR.